src/blog/2024/02/postgresql-with-node-red.md

```diff
@@ -0,0 +1,233 @@
+
+# Creating Table
+
+In this section, we will create a table in our database to store product data.
+
+1. Drag an Inject node onto canvas, and keep it unchanged.
+2. Click on the PostgreSQL node that we have added previously, and paste the following SQL command into the query input field. (I have added comments for your understanding of SQL commands)
+
+```sql
+-- Create a table named product_data if it does not already exist
+CREATE TABLE IF NOT EXISTS product_data (
+  -- Define a column named id as a bigint type, which serves as the primary key
+  id bigint PRIMARY KEY,
+  
+  -- Define a column named name to store product names as variable-length character strings with a maximum length of 100 characters, ensuring it's not null
+  name varchar(100) NOT NULL,
+  
+  -- Define a column named price to store product prices, ensuring it's not null
+  price int NOT NULL,
+  
+  -- Define a column named stock to store product stock levels, ensuring it's not null
+  stock int NOT NULL
+);
+```
+!["Creating table for product data"](./images/postgresql_with_nodered_create_table.png "Creating table for product data")
+
+3. Connect the inject node’s output to the PostgreSQL node’s input.
+
+# Installing Dashboard 2.0 
+
+1. Install Dashboard 2.0. Follow these [instructions](https://dashboard.flowfuse.com/getting-started.html) to install.
+2. Create your first group, page, theme, and base.
+
+# Inserting Product Data into the Database
+In this section, we will add a Form interface that will enable us to obtain product data that we need to insert into the database. Moreover, we will use the PostgreSQL node to interact with the database.
+
+!["Adding form to insert data"](./images/postgresql_with_node-red_form1.png "Adding form to insert data")
+1. Drag a Ui-Form widget onto the canvas and select the created group.
+2. Add an element for all required input data in the form widget and give it a name, label, and select type, I have selected 'number' as a type for 'product id', 'price', and 'stock', and 'text' for 'name', but feel free to adjust according to your preference and data requirements.
+3. Drag the function node onto Canvas and paste the following script.
+
+```javascript
+// Destructure the properties from msg.payload (obtained data by using form)
+const { id, name, price, stock } = msg.payload;
+// Create an array containing id, name, price, and stock
+// The order of the array items in msg.params will correspond to the placeholders in the SQL query
+// For example, $1 will be replaced by the value of id, $2 will be replaced by the value of name, and so on
+msg.params = [id, name, price, stock];
+return msg;
+```
+
+4. Drag a PostgreSQL node onto the Canvas and click on that node and paste the following SQL command into the query input field
+
+```sql
+-- This is an SQL INSERT statement used to add data into the product_data table.
+INSERT INTO product_data (id, name, price, stock)
```

Comment:
If you have a self-incrementing (SERIAL) ID field, you don't need to insert this.

Reply:
Okay ZJ, I will update the image, script, and SQL command accordingly.